module/zfs/spa_misc.c

```diff
@@ -2019,6 +2019,236 @@ spa_dirty_data(spa_t *spa)
+ * SPA Import Progress Routines
+ * ==========================================================================
+ */
+
+typedef struct spa_import_progress {
+	uint64_t		pool_guid;	/* unique # for updates */
+	char			*pool_name;
+	spa_load_state_t	spa_load_state;
+	uint64_t		check_start;	/* UTC time MMP check start */
+	uint64_t		check_end;	/* UTC time MMP check end */
+	uint64_t		spa_load_max_txg;	/* rewind txg */
+	procfs_list_node_t	smh_node;
+} spa_import_progress_t;
+
+spa_history_list_t *spa_import_progress_list = NULL;
+unsigned int spa_import_progress_list_size = 16;
+
+static int
+spa_import_progress_show_header(struct seq_file *f)
+{
+	seq_printf(f, "%-20s %-14s %-12s %-12s %-12s %s\n", "pool_guid",
+	    "load_state", "check_start", "check_end", "max_txg",
+	    "pool_name");
+	return (0);
+}
+
+static int
+spa_import_progress_show(struct seq_file *f, void *data)
+{
+	spa_import_progress_t *sip = (spa_import_progress_t *)data;
+
+	seq_printf(f, "%-20llu %-14llu %-12llu %-12llu %-12llu %s\n",
+	    (u_longlong_t)sip->pool_guid, (u_longlong_t)sip->spa_load_state,
+	    (u_longlong_t)sip->check_start, (longlong_t)sip->check_end,
+	    (u_longlong_t)sip->spa_load_max_txg,
+	    (sip->pool_name ? sip->pool_name : "-"));
+
+	return (0);
+}
+
+/* Remove oldest elements from list until there are no more than 'size' left */
+static void
+spa_import_progress_truncate(spa_history_list_t *shl, unsigned int size)
+{
+	spa_import_progress_t *sip;
+	while (shl->size > size) {
+		sip = list_remove_head(&shl->procfs_list.pl_list);
+		if (sip->pool_name)
+			strfree(sip->pool_name);
+		kmem_free(sip, sizeof (spa_import_progress_t));
+		shl->size--;
+	}
+
+	if (size == 0)
+		ASSERT(list_is_empty(&shl->procfs_list.pl_list));
+
+}
+
+static int
+spa_import_progress_clear(procfs_list_t *procfs_list)
+{
+	spa_history_list_t *shl = procfs_list->pl_private;
+	mutex_enter(&procfs_list->pl_lock);
+	spa_import_progress_truncate(shl, 0);
+	mutex_exit(&procfs_list->pl_lock);
+	return (0);
+}
+
+static void
+spa_import_progress_init(void)
+{
+	spa_import_progress_list = kmem_zalloc(sizeof (spa_history_list_t),
+	    KM_SLEEP);
+
+	spa_import_progress_list->size = 0;
+
+	spa_import_progress_list->procfs_list.pl_private =
+	    spa_import_progress_list;
+
+	procfs_list_install("zfs",
+	    "import_progress",
+	    0644,
+	    &spa_import_progress_list->procfs_list,
+	    spa_import_progress_show,
+	    spa_import_progress_show_header,
+	    spa_import_progress_clear,
+	    offsetof(spa_import_progress_t, smh_node));
+}
+
+static void
+spa_import_progress_destroy(void)
+{
+	spa_history_list_t *shl = spa_import_progress_list;
+	procfs_list_uninstall(&shl->procfs_list);
+	spa_import_progress_truncate(shl, 0);
+	kmem_free(shl, sizeof (spa_history_list_t));
+	procfs_list_destroy(&shl->procfs_list);
+}
+
+/*
+ * For update methods, it is important that we start search at the tail of the
+ * list where new records are inserted, so this is normally an O(1) operation.
+ */
```

Comment:
I'm not sure this comment is true since we're matching on the `pool_guid`.  If multiple long running concurrent imports are allowed the one we're interested in could appear anywhere in the list.

Reply:
Agreed.